lib/with_win32.dart

```diff
@@ -0,0 +1,155 @@
+import 'dart:ffi';
+import 'package:ffi/ffi.dart';
+import 'win32.dart' as win32;
+
+export 'win32.dart'
+    hide
+        CreateWindowEx,
+        DispatchMessage,
+        DrawText,
+        GetClientRect,
+        GetMessage,
+        RegisterClass,
+        TranslateMessage;
+
+T withMSG<T>(T Function(win32.MSG) func) {
+  final msg = win32.MSG.allocate();
+  T t;
+  try {
+    t = func(msg);
+  } finally {
+    free(msg.addressOf);
+  }
+  return t;
+}
+
+T withPAINTSTRUCT<T>(T Function(win32.PAINTSTRUCT) func) {
+  final ps = win32.PAINTSTRUCT.allocate();
+  T t;
+  try {
+    t = func(ps);
+  } finally {
+    free(ps.addressOf);
+  }
+  return t;
+}
+
+T withRECT<T>(T Function(win32.RECT) func) {
+  final rect = win32.RECT.allocate();
+  T t;
+  try {
+    t = func(rect);
+  } finally {
+    free(rect.addressOf);
+  }
+  return t;
+}
+
+T withTEXT<T>(String text, T Function(Pointer<Utf16>) func) {
+  final utf16Text = win32.TEXT(text);
+  T t;
+  try {
+    t = func(utf16Text);
+  } finally {
+    free(utf16Text);
+  }
+  return t;
+}
+
+T withWNDCLASS<T>(T Function(win32.WNDCLASS) func) {
+  final wc = win32.WNDCLASS.allocate();
+  T t;
+  try {
+    t = func(wc);
+  } finally {
+    free(wc.addressOf);
+  }
+  return t;
+}
+
+T BeginPaint<T>(int hWnd, T Function(int hdc) func) => withPAINTSTRUCT((ps) {
+      final hdc = win32.BeginPaint(hWnd, ps.addressOf);
+      T t;
+      try {
+        t = func(hdc);
+      } finally {
+        win32.EndPaint(hWnd, ps.addressOf);
+      }
+      return t;
+    });
+
+int CreateWindowEx(
+        int dwExStyle,
+        String className,
+        String windowName,
+        int dwStyle,
+        int X,
+        int Y,
+        int nWidth,
+        int nHeight,
+        int hWndParent,
+        int hMenu,
+        int hInstance,
+        Pointer<Void> lpParam) =>
+    withTEXT(
+      className,
+      (className) => withTEXT(
+        windowName,
+        (windowName) {
+          var hWnd = win32.CreateWindowEx(
+            dwExStyle,
+            className,
+            windowName,
+            dwStyle,
+            X,
+            Y,
+            nWidth,
+            nHeight,
+            hWndParent,
+            hMenu,
+            hInstance,
+            lpParam,
+          );
+
+          if (hWnd == 0) {
+            final error = win32.GetLastError();
+            throw win32.WindowsException(win32.HRESULT_FROM_WIN32(error));
+          }
+
+          return hWnd;
+        },
+      ),
+    );
+
+int DispatchMessage(win32.MSG msg) => win32.DispatchMessage(msg.addressOf);
+
+int DrawText(int hdc, String text, int cchText, win32.RECT rect, int format) =>
+    withTEXT(
+        text,
+        (utf16text) =>
+            win32.DrawText(hdc, utf16text, cchText, rect.addressOf, format));
+
+T GetClientRect<T>(int hWnd, T Function(win32.RECT rect) func) =>
+    withRECT((rect) {
+      var result = win32.GetClientRect(hWnd, rect.addressOf);
+      if (result != 0) {
+        return func(rect);
```

Comment:
Does the `finally` block of a `try/finally` execute even when there's a `return`? I could only find reference to it firing regardless of whether an exception is thrown or not.

Reply:
I expect so, given that is how `try/finally` works in other languages. Time to ask a language lawyer?